mqtt/mqtt-bridge/src/client.rs

```diff
@@ -32,84 +40,189 @@ impl ShutdownHandle {
+                    } else {
+                        futures_util::future::Either::Right(futures_util::future::ok(None))
+                    };
+
+                    let token = if let Some(ref ts) = token_source {
+                        futures_util::future::Either::Left(ts.get(&expiry))
+                    } else {
+                        futures_util::future::Either::Right(futures_util::future::ok(None))
+                    };
+
+                    let io = TcpStream::connect(host);
+
+                    let (server_root_certificate, password, stream) =
+                        futures_util::future::try_join3(server_root_certificate, token, io)
+                            .await
+                            .map_err(|err| {
+                                Error::new(ErrorKind::Other, format!("failed to connect: {}", err))
+                            })?;
+
+                    let mut builder = TlsConnector::builder();
+
+                    if let Some(trust_bundle) = server_root_certificate {
+                        let certs = X509::stack_from_pem(trust_bundle.as_bytes())
+                            .unwrap()
+                            .into_iter()
+                            .map(|cert| Certificate::from_der(&cert.to_der().unwrap()).unwrap());
+
+                        for cert in certs {
+                            builder.add_root_certificate(cert);
+                        }
+                    }
+
+                    let connector = builder.build().map_err(|err| {
+                        Error::new(
+                            ErrorKind::Other,
+                            format!("could not create TLS connector: {}", err),
+                        )
+                    })?;
+                    let connector: tokio_tls::TlsConnector = connector.into();
+
+                    let io = connector.connect(&address, stream).await;
+
+                    debug!("Tls connection {:?} for {:?}", io, address);
+
+                    io.map_or_else(
+                        |e| Err(Error::new(ErrorKind::Other, e)),
+                        |io| {
+                            let stream: Pin<Box<dyn BridgeIo>> = Box::pin(io);
+                            Ok((stream, password))
+                        },
+                    )
+                })
+            }
+        }
     }
 }
 
 pub struct MqttClient<T>
 where
     T: EventHandler,
 {
-    client: Client<TcpConnection<SasTokenSource>>,
+    client_id: String,
+    username: Option<String>,
+    io_source: BridgeIoSource,
+    keep_alive: Duration,
+    client: Client<BridgeIoSource>,
     event_handler: T,
 }
 
 impl<T: EventHandler> MqttClient<T> {
     pub fn new(
         address: &str,
+        port: Option<String>,
         keep_alive: Duration,
-        clean_session: bool,
+        _clean_session: bool,
         event_handler: T,
         connection_credentials: &Credentials,
+        secure: bool,
```

Comment:
there should be a bool at some point to configure if it's tls or not so if I add 2 ctors I will have to check the value and call tls() or tcp(), so it moves the decision up

Reply:
also, it can be mqtt over websockets, right?

Reply:
I vote for @dmolokanov suggestion. @ancaantochi, I understand your point that we would need to make the decision up in the stack, but I tend to avoid such bools as params and I believe having separate method will lead to a better design.

Reply:
> also, it can be mqtt over websockets, right?
Not implemented, I think it might be after public preview